File: src/test/java/com/morak/performancetracker/ContextTypePropertiesTest.java
package com.morak.performancetracker;

import static org.assertj.core.api.Assertions.assertThat;

import com.morak.performancetracker.context.ContextManager;
import com.morak.performancetracker.context.EndpointContextManager;
import com.morak.performancetracker.context.MethodContextManager;
import java.util.List;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.TestPropertySource;

@SpringBootTest
@PerformanceTracker
public class ContextTypePropertiesTest {

    @Nested
    class Context_Type에_아무것도_없는경우 {

        @Autowired
        private List<ContextManager> managers;

        @Test
        void 모든_Type이_등록된다() {
            assertThat(managers).hasOnlyElementsOfTypes(MethodContextManager.class, EndpointContextManager.class);
        }
    }

    @Nested
    @TestPropertySource(properties = "com.morak.performance-tracker.context.type = method")
    class Context_Type에_Method만_있는_경우 {

        @Autowired
        private List<ContextManager> managers;

        @Test
        void Method_타입만_등록된다() {
            assertThat(managers).hasExactlyElementsOfTypes(MethodContextManager.class);
        }
    }

    @Nested
    @TestPropertySource(properties = "com.morak.performance-tracker.context.type = endpoint")
    class Context_Type에_Endpoint만_있는_경우 {

        @Autowired
        private List<ContextManager> managers;

        @Test
        void Endpoint_타입만_등록된다() {
            assertThat(managers).hasExactlyElementsOfTypes(EndpointContextManager.class);
        }
    }

    @Nested
    @TestPropertySource(properties = "com.morak.performance-tracker.context.type = method, endpoint")
    class Context_Type에_여러개를_설정하는_경우 {

        @Autowired
        private List<ContextManager> managers;

        @Test
        void 설정한_모든_타입이_등록된다() {
            assertThat(managers).hasOnlyElementsOfTypes(MethodContextManager.class, EndpointContextManager.class);
        }
    }
}

File: src/test/kotlin/com/morak/performancetracker/ContextTypePropertiesTest.kt
package com.morak.performancetracker

import com.morak.performancetracker.context.ContextManager
import com.morak.performancetracker.context.EndpointContextManager
import com.morak.performancetracker.context.MethodContextManager
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.TestPropertySource

@SpringBootTest
@PerformanceTracker
class ContextTypePropertiesTest {
    @Nested
    inner class `Context Type에 아무것도 없는경우` {
        @Autowired
        private lateinit var managers: List<ContextManager>

        @Test
        fun `모든 Type이 등록된다`() {
            assertThat(managers).hasOnlyElementsOfTypes(
                MethodContextManager::class.java,
                EndpointContextManager::class.java
            )
        }
    }

    @Nested
    @TestPropertySource(properties = ["com.morak.performance-tracker.context.type = method"])
    inner class `Context Type에 Method만 있는 경우` {
        @Autowired
        private lateinit var managers: List<ContextManager>

        @Test
        fun `Method 타입만 등록된다`() {
            assertThat(managers).hasExactlyElementsOfTypes(MethodContextManager::class.java)
        }
    }

    @Nested
    @TestPropertySource(properties = ["com.morak.performance-tracker.context.type = endpoint"])
    inner class `ContextType에 Endpoint만 있는 경우` {
        @Autowired
        private lateinit var managers: List<ContextManager>

        @Test
        fun `Endpoint 타입만 등록된다`() {
            assertThat(managers).hasExactlyElementsOfTypes(EndpointContextManager::class.java)
        }
    }

    @Nested
    @TestPropertySource(properties = ["com.morak.performance-tracker.context.type = method, endpoint"])
    inner class `ContextType에 여러개를 설정하는 경우` {
        @Autowired
        private lateinit var managers: List<ContextManager>
        @Test
        fun `설정한 모든 타입이 등록된다`() {
            println("managers = ${managers}")
            assertThat(managers).hasOnlyElementsOfTypes(
                MethodContextManager::class.java,
                EndpointContextManager::class.java
            )
        }
    }
}